File: agent/feedback.py
"""Computes and persists feedback for yesterday's brief.
Runs at the start of each agent run — compares yesterday's tasks against today's
to record what was resolved, what's still open, and how long carried tasks took.
Also reads user reactions (👍/👎) written by the plugin from _generated/feedback/reactions/.
"""
import json
from pathlib import Path
from datetime import date, timedelta

from .carry_detector import _fuzzy_match


def compute_feedback(today_tasks: dict, config: dict) -> dict | None:
    vault_path = Path(config["vault_path"])
    output_dir = vault_path / config["output_dir"]
    feedback_dir = vault_path / config["feedback_dir"]

    yesterday = (date.today() - timedelta(days=1)).isoformat()
    yesterday_brief_path = output_dir / f"{yesterday}.json"

    if not yesterday_brief_path.exists():
        return None

    yesterday_brief = json.loads(yesterday_brief_path.read_text(encoding="utf-8"))

    today_all_texts = [
        t["text"] if isinstance(t, dict) else t
        for cat in ("red_alert", "regular")
        for t in today_tasks.get(cat, [])
    ]

    yesterday_all_tasks = (
        yesterday_brief.get("tasks", {}).get("red_alert", [])
        + yesterday_brief.get("tasks", {}).get("regular", [])
    )

    resolved = []
    still_open = []
    days_carried = {}

    for task in yesterday_all_tasks:
        task_text = task["text"]
        still_present = any(_fuzzy_match(task_text, t) for t in today_all_texts)
        if still_present:
            still_open.append(task_text)
        else:
            resolved.append(task_text)
            if task.get("carried_from"):
                days = (date.today() - date.fromisoformat(task["carried_from"])).days
                days_carried[task_text] = days

    suggestions = yesterday_brief.get("suggestions", [])
    suggestions_resolved = []
    for s in suggestions:
        if s and s.get("text"):
            hit = any(
                _fuzzy_match(r, word)
                for r in resolved
                for word in s["text"].split() if len(word) > 10
            )
            if hit:
                suggestions_resolved.append(s["text"])

    reactions_dir = feedback_dir / "reactions"
    reaction_file = reactions_dir / f"{yesterday}.json"
    user_reactions = None
    if reaction_file.exists():
        user_reactions = json.loads(reaction_file.read_text(encoding="utf-8"))

    feedback = {
        "date": yesterday,
        "picks": {
            "tactical_rules_shown": yesterday_brief.get("tactical_rules", []),
            "identity_rules_shown": yesterday_brief.get("identity", {}).get("rules", []),
            "hobby_tasks_shown": yesterday_brief.get("hobby_tasks", []),
            "suggestions": suggestions,
        },
        "outcomes": {
            "suggestions_resolved": suggestions_resolved,
            "carried_tasks_resolved": resolved,
            "carried_tasks_still_open": still_open,
            "days_carried_before_resolve": days_carried,
            "new_tasks_added": len(today_all_texts) - len(still_open),
        },
        "user_reactions": user_reactions,
    }

    feedback_dir.mkdir(parents=True, exist_ok=True)
    feedback_path = feedback_dir / f"{yesterday}.json"
    feedback_path.write_text(json.dumps(feedback, indent=2), encoding="utf-8")

    return feedback

File: src/agent/feedback.ts
import { App } from "obsidian";
import type { MorningOSSettings } from "../settings";
import { fuzzyMatch } from "./carry-detector";
import { findMostRecentHistory } from "./history";
import { findMostRecentBrief } from "./find-recent-brief";

export async function computeFeedback(
  todayTasks: { red_alert: { text: string; status_completion?: string; done?: boolean }[]; regular: { text: string; status_completion?: string; done?: boolean }[] },
  todayStr: string,
  app: App,
  settings: MorningOSSettings
): Promise<void> {
  const [historyFound, briefFound] = await Promise.all([
    findMostRecentHistory(todayStr, app, settings.carryLookbackDays),
    findMostRecentBrief(todayStr, app, settings),
  ]);
  if (!historyFound && !briefFound) return;

  const yesterdayStr = historyFound?.date ?? briefFound?.date ?? "";
  const yesterdayBrief = briefFound?.brief;
  const yesterdaySnapshot = historyFound?.snapshot;

  const todayAllTexts: string[] = [
    ...todayTasks.red_alert.filter(t => t.status_completion !== "done" && !t.done).map(t => t.text),
    ...todayTasks.regular.filter(t => t.status_completion !== "done" && !t.done).map(t => t.text),
  ];

  const yesterdayAllTasks = [
    ...(yesterdaySnapshot?.tasks?.red_alert ?? []),
    ...(yesterdaySnapshot?.tasks?.regular ?? []),
  ];

  const resolved: string[] = [];
  const stillOpen: string[] = [];
  const daysCarried: Record<string, number> = {};

  const todayDate = new Date(todayStr + "T12:00:00");

  for (const task of yesterdayAllTasks) {
    const taskText = task.text;
    const stillPresent = todayAllTexts.some(t => fuzzyMatch(taskText, t));
    if (stillPresent) {
      stillOpen.push(taskText);
    } else {
      resolved.push(taskText);
      if (task.carried_from) {
        const carriedDate = new Date(task.carried_from + "T12:00:00");
        const days = Math.round((todayDate.getTime() - carriedDate.getTime()) / 86400000);
        daysCarried[taskText] = days;
      }
    }
  }

  const suggestions: { text: string; source: string }[] = yesterdayBrief?.suggestions ?? [];
  const suggestionsResolved: string[] = [];
  for (const s of suggestions) {
    if (s?.text) {
      const hit = resolved.some(r =>
        s.text.split(" ").some(word => word.length > 10 && fuzzyMatch(r, word))
      );
      if (hit) suggestionsResolved.push(s.text);
    }
  }

  const reactionsPath = `${settings.feedbackDir}/reactions/${yesterdayStr}.json`;
  let userReactions: unknown = null;
  const reactionsExist = await app.vault.adapter.exists(reactionsPath);
  if (reactionsExist) {
    try {
      const raw = await app.vault.adapter.read(reactionsPath);
      userReactions = JSON.parse(raw);
    } catch {
      // ignore
    }
  }

  const feedback = {
    date: yesterdayStr,
    picks: {
      tactical_rules_shown: yesterdayBrief?.tactical_rules ?? [],
      identity_rules_shown: yesterdayBrief?.identity?.rules ?? [],
      hobby_tasks_shown: yesterdayBrief?.hobby_tasks ?? [],
      suggestions,
    },
    outcomes: {
      suggestions_resolved: suggestionsResolved,
      carried_tasks_resolved: resolved,
      carried_tasks_still_open: stillOpen,
      days_carried_before_resolve: daysCarried,
      new_tasks_added: todayAllTexts.length - stillOpen.length,
    },
    user_reactions: userReactions,
  };

  await app.vault.adapter.mkdir(settings.feedbackDir);
  await app.vault.adapter.write(
    `${settings.feedbackDir}/${yesterdayStr}.json`,
    JSON.stringify(feedback, null, 2)
  );
}
